### ISSUE TITLE
AWX and ANSIBLE_ prefixed environment variables - support a 2nd set of env vars? ansible vars?

### ISSUE THREAD

**Issue by morco** [ansible-collections/community.hashi_vault#49]
##### SUMMARY
Nothing against your new environment variable naming scheme (https://github.com/ansible-collections/community.hashi_vault/issues/10#issue-754561671), but would you consider supporting a second set of environment variables which does not use the `ANSIBLE_` prefix but maybe something like `ANS_` or similar?

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

hashivault lookup plugin

##### ADDITIONAL INFORMATION

The problem we have with the `ANSIBLE_` prefix is that `AWX` does not allow user credentials to induce environment variables starting with this prefix.

> When creating a new credential type, you are responsible for avoiding collisions in the extra_vars, env, and file namespaces. Also, avoid environment variable or extra variable names that start with ANSIBLE_ because they are reserved.

[SOURCE](https://docs.ansible.com/ansible-tower/latest/html/userguide/credential_types.html)

**Comment by briantist** [comment]
Hi @morco thanks for raising this. Admittedly I don't have any experience with AWX so I really appreciate the perspective. Since this is referencing "credentials" I assume this is a way to deal with values that are sensitive/secret? As in the env vars this would apply to are most likely `ANSIBLE_HASHI_VAULT_TOKEN`, `ANSIBLE_HASHI_VAULT_SECRET_ID`, etc.?

Another thing I'm curious about: you mentioned AWX does not allow them, but from the wording, it seems like they are discouraging it to avoid inadvertently overriding an existing value, and the responsibility is on the caller.

 My read of that is that they are recommending not to use _custom_ names beginning with `ANSIBLE_` to avoid conflicts with pre-defined variables and uses. Based on that, I feel like setting these would be an appropriate use of the feature. 

Put another way: I think this plugin's use of the prefix would be considered one of the aforementioned "reserved" uses, and you'd be setting/overriding the value intentionally (nearly every `env:` entry you'll find for other plugins uses the prefix). Again, I have low context around AWX so please correct me if I've misunderstood any of these points.

Another thing I'm in early stages of thinking about is adding `vars:` entries to the options in the plugins. That would be another way of specifying them, via setting variables in Ansible, which it sounds like is also usable by this credentials feature (but those would likely also use the `ansible_` prefix).

**Comment by morco** [comment]
> 
> 
> Hi @morco thanks for raising this. Admittedly I don't have any experience with AWX so I really appreciate the perspective. Since this is referencing "credentials" I assume this is a way to deal with values that are sensitive/secret? As in the env vars this would apply to are most likely `ANSIBLE_HASHI_VAULT_TOKEN`, `ANSIBLE_HASHI_VAULT_SECRET_ID`, etc.?
>

Correct. We actually have all auth related params there, not only the secret ones (url, authmethod, ...). This makes it imho the ideal method for managing authing inside awx. You create the credential once and just have to reference it in job configurations (aka playbook runs) while in your role/playbook you dont have to specify how authing is handled every time it is needed like when using the hashivault lookup. We used it quite successfully with the old `VAULT_` variables.

> 
> Another thing I'm curious about: you mentioned AWX does not allow them, but from the wording, it seems like they are discouraging it to avoid inadvertently overriding an existing value, and the responsibility is on the caller.
> 
> My read of that is that they are recommending not to use _custom_ names beginning with `ANSIBLE_` to avoid conflicts with pre-defined variables and uses. Based on that, I feel like setting these would be an appropriate use of the feature.
> 
> Put another way: I think this plugin's use of the prefix would be considered one of the aforementioned "reserved" uses, and you'd be setting/overriding the value intentionally (nearly every `env:` entry you'll find for other plugins uses the prefix). Again, I have low context around AWX so please correct me if I've misunderstood any of these points.
>

I agree with you that there are legimate cases like this for this prefix and the final call on this should have the person which manages the awx instance, but unfortunately it seems to be a hardcoded no-no:

![awx_badenv_prefix](https://user-images.githubusercontent.com/259572/105207020-4d24e000-5b47-11eb-8621-8a3bea7b075e.png)

To make this clear, beside the error message showing the save button simply does nothing in this case. So ultimately one can totally argue that this is an issue of awx, not of plugins. I also must admit that our server is not running on the latest version of awx as we currently have yet to resolve upgrade/migration issues, so it might be totally possible that this behaviour has changed in newer versions. The up to date docu I cited above makes me still a bit sceptical about that.

>
> Another thing I'm in early stages of thinking about is adding `vars:` entries to the options in the plugins. That would be another way of specifying them, via setting variables in Ansible, which it sounds like is also usable by this credentials feature (but those would likely also use the `ansible_` prefix).
>

I like that a lot. I have tested it and interestingly our awx has no issue with inducing variables starting with the `ansible_` prefix. Even if it would not work directly variables would give one the flexibility to remap names as needed by roles/actions/set_fact etc.

So is it more a matter of time or are you still deliberating if to do this at all? I personally use it very often in the modules/actions I write. I find it quite useful and dont see a good reason for not using vars, but than again I dont have seen any "upstream" modules yet supporting defaulting by ansible variables, so maybe there are good reasons.

**Comment by briantist** [comment]
> but than again I dont have seen any "upstream" modules yet supporting defaulting by ansible variables, so maybe there are good reasons.

The main examples you'll see are actually connection plugins, via variables like `ansible_ssh_whatever`, so even though it's been far less common in other types, there is in fact built-in support for `vars:` being specified in the argspec (in documentation) that gets processed by config manager, same as for `env:` and `ini:` entries. There is a recent issue I saw though in the case of templating `vars:` in a task that are then interpreted by a plugin using `vars:`: https://github.com/ansible/ansible/issues/73268 (but there are ways to workaround it and that part needs to be solved in core anyway).

> So is it more a matter of time or are you still deliberating if to do this at all?

Before adding those, I want to do some other work to centralize the processing of auth methods into a module_util. Right now the `hashi_vault` plugin is mostly dedicated to process auth methods, and as a result the testing keeps getting more and more complex. 

I want to get code handling auth methods in one place, with its own unit and integration tests, and simplify this plugin a lot.

Which leads me to this:
> dont have to specify how authing is handled every time it is needed like when using the hashivault lookup

And I'm so glad you brought this up because I have the same issue where I use this plugin in production as well. The work I mentioned above, also leads to being able to make new plugins (and maybe modules) much more easily, and that lends itself well to workflows where you can auth once, and then use simpler plugins that do one thing well, and can just accept a token.

This stuff isn't completely solidified or anything yet, but that's the direction I'm looking to go and where I'm spending free time on at the moment.

So to get back to the question of adding `vars:` entries for the existing plugin:  yes it _could_ be done sooner than that, but I'd like to try to get the basics of refactoring done: namely, moving code centrally and adding unit/integration tests for it.

Depending on the trajectory of that work, I may move toward adding vars sooner than that. It's nice to know the feature would see use other than from me, for sure!

I'm going to see if anyone in the IRC channel might know more about the immediate issue, maybe there's a setting that can allow you to override the env vars...

**Comment by morco** [comment]
>
> So to get back to the question of adding `vars:` entries for the existing plugin: yes it _could_ be done sooner than that, but I'd like to try to get the basics of refactoring done: namely, moving code centrally and adding unit/integration tests for it.
> 
> Depending on the trajectory of that work, I may move toward adding vars sooner than that. It's nice to know the feature would see use other than from me, for sure!
> 
> I'm going to see if anyone in the IRC channel might know more about the immediate issue, maybe there's a setting that can allow you to override the env vars...

Thanks. I already tried some things, but none of them succesful. Most of the ansible environment manipulating stuff like the `environment:` keyword is tailored for remote targets and modules, not lookup's on the controller (even when target is `localhost`). Another thing I tried is to manipulate `os.environ` in an action plugin. This should have worked as far as I can tell if ansible is run as one big python process. As it did not I guess some subprocessing is involved.
But dont worry about me, I have an usable workaround for the moment. I basically wrap/subclass the `community.hashivault` lookup plugin. This is obviously less than ideal but works for now.

This was for me more a question of will it work eventually than I need this ASAP. So no pressure, take your time, it sounds like a good plan :)

**Comment by briantist** [comment]
Thanks @morco I really appreciate the discussion, I was actually going to suggest a wrapper lookup as that's exactly what I'm doing in $DAYJOB in order to use vars :)

I haven't really gotten much of a response in IRC either, but if you want to try: http://webchat.freenode.net/?channels=ansible-awx

It might also be worth raising an issue on the AWX github, I looked back and I did one thing of interest: https://github.com/ansible/awx/pull/2363

This shows that it didn't use to enforce this via not being able to save as shown in your screenshot, but even when that was the case, it didn't work (the `ANSIBLE_` prefixed env vars were silently skipped).

You mentioned it didn't prevent you from saving vars prefixed with `ansible_`; I wonder if that actually works or is an unfixed case of them being silently skipped. If that's the case then even the plan to add vars to the options won't work, I think.

Anyway, as I make progress on some of these changes that might be of interest to you, I may tag you to take a look in case you have time. Thanks again!

**Comment by erinn** [comment]
I'll add to this, that as things currently stand regard ANSIBLE_HASHI* variables and AWX things simply no longer work with the latest release. 

Though other variables are deprecated we are still able to use older environment variable names like: 
VAULT_ROLE_ID
VAULT_SECRET_ID

with the current release, VAULT_ADDR however, does NOT seem to work. Thus we have ended up hacking around this issue using the following psuedo code:

`"{{ lookup('community.hashi_vault.hashi_vault', secret=secret_stuff  url=lookup('env', 'VAULT_ADDR') auth_method=approle') }}"`

It works, but this is a bit less than ideal. Again this is with the latest release of hashi_vault.

I understand you are in the middle of rearchitecting things, but my impression is this is a bit broken and going against what is documented in terms of deprecations, I believe, if I understand things right that though VAULT_ADDR is deprecated it should still work. I'd recommend not forcing a move to ANSIBLE_HASHI* variables until such time as this is all sorted. 

Much appreciative for your work.

**Comment by briantist** [comment]
Hi @erinn ! `VAULT_ADDR`'s breaking change was documented in the release notes for `community.hashi_vault` 1.0.0: https://github.com/ansible-collections/community.hashi_vault/releases/tag/1.0.0

You may use 0.2.0 if you still need the value of `VAULT_ADDR` to override other sources, but I recommend your workaround instead if you are looking to make an intentional choice about the value of `VAULT_ADDR` being used above other sources.


`VAULT_ADDR` falls into the category of "low-precedence" env vars and its use for that is not deprecated, so it definitely should work _if there are no other sources_. If you're seeing something different definitely let me know.

Since the change was only to _lower_ the precedence of `VAULT_ADDR` compared to the value set in INI or `ANSIBLE_HASHI_VAULT_ADDR`, if it has stopped working, I recommend checking your `ansible.cfg` file to see if it may have been defined in there, or if there's any other place the vault address may have been defined in your environment.

---

Also you may have just left it out of the example for brevity, but just in case, your workaround could be eased slightly by doing the controller lookup once, via `vars:` or `set_fact:` and then using a var in all the `lookup` calls:
```yaml
- hosts: localhost
  vars:
    vurl: "{{ lookup('env', 'VAULT_ADDR') }}"
  tasks:
    - debug:
        msg: "{{ lookup('community.hashi_vault.hashi_vault', 'secret=secret_stuff', url=vurl, auth_method='approle') }}"
        # more vault calls
```

**Comment by elcomtik** [comment]
> `VAULT_ADDR` falls into the category of "low-precedence" env vars and its use for that is not deprecated, so it definitely should work _if there are no other sources_. If you're seeing something different definitely let me know.

Hi @briantist, I experience the same issue. I use  AWX 17.0.1, which doesn't allow ANSIBLE_ prefixed env vars, so I'm forced to use the old one VAULT_ADDR. I will try to reach AWX devs and consult this behavior. Meantime I think here in hashi_lookup is an issue, which doesn't work as we would like.

I have not specified vault url in ansible.cfg, I have specified only env VAULT_ADDR. When I run playbook it uses the default "http://127.0.0.1:8200" instead of the value specified by env VAULT_ADDR.

Shouldn't the default option have the lowest priority?

BTW: Why set any default for url? Wouldn't by better to make it required and fail quickly if it is not set? Why we suppose, that "http://127.0.0.1:8200" is better than none?

**Comment by briantist** [comment]
Hi @elcomtik a similar issue was raised recently (#60) so I'm going to look into if `VAULT_ADDR` is being skipped in that ticket (it definitely is not the desired behavior).

As for your question of the default: I can't speak to the original motivations as I started maintaining the plugin after its initial creation, but it was probably motivated by the fact that that is the default used by the vault CLI for example, and that address is useful if you are running a vault dev server on the local machine, or Vault Agent.

**Comment by elcomtik** [comment]
At first thanks for your work on this module, I have heavy use of it and I'm glad it is further maintained. 

> Hi @elcomtik a similar issue was raised recently (#60) so I'm going to look into if `VAULT_ADDR` is being skipped in that ticket (it definitely is not the desired behavior).

That ticket describes exactly that what I have experienced.

> As for your question of the default: I can't speak to the original motivations as I started maintaining the plugin after its initial creation, but it was probably motivated by the fact that that is the default used by the vault CLI for example, and that address is useful if you are running a vault dev server on the local machine, or Vault Agent.

I personally think such default is dangerous, this should be handled in the environment variables.

**Comment by briantist** [comment]
> At first thanks for your work on this module, I have heavy use of it and I'm glad it is further maintained.

Thank you much appreciated, I think there is a lot of exciting stuff coming.

> > As for your question of the default: I can't speak to the original motivations as I started maintaining the plugin after its initial creation, but it was probably motivated by the fact that that is the default used by the vault CLI for example, and that address is useful if you are running a vault dev server on the local machine, or Vault Agent.
> 
> I personally think such default is dangerous, this should be handled in the environment variables.

It would be a breaking change for this plugin to have no default, but I will consider it both for this one and any new plugins/modules.

I would normally agree with you, but it is harder make that case when Vault's own tools have a default address, and so it's less clear what users' expectations are around that.

**Comment by elcomtik** [comment]
> It would be a breaking change for this plugin to have no default, but I will consider it both for this one and any new plugins/modules.

Yeah, nobody likes breaking changes ;)

> I would normally agree with you, but it is harder make that case when Vault's own tools have a default address, and so it's less clear what users' expectations are around that.

In my opinion, ansible should not lean to defaults of any CLI tool related to its modules.

**Comment by pilou-** [comment]
> BTW: Why set any default for url? Wouldn't by better to make it required and fail quickly if it is not set? Why we suppose, that "http://127.0.0.1:8200" is better than none?

The `hashi_vault` lookup plugin relies on [hvac](https://github.com/hvac/hvac/commit/322460d6f752057ac76a2c530dd53886faf7fd27#diff-51f67bc9cea7748a338f0250807d36fd9443bedd4f2e3f953089cc006542bda9R6) library which uses [`http://localhost:8200` as a default value](https://github.com/hvac/hvac/commit/322460d6f752057ac76a2c530dd53886faf7fd27#diff-51f67bc9cea7748a338f0250807d36fd9443bedd4f2e3f953089cc006542bda9R6) since a long time (before [`VAULT_ADDR` was handled](https://github.com/hvac/hvac/commit/591f6d1fa8601bce4ecd641036c2bb58de22f3e5)).
The `vault` executable uses [another default value](https://github.com/hashicorp/vault/blob/22b486b651b8956d32fb24e77cef4050df7094b6/api/client.go#L143): `https://127.0.0.1:8200` and [`VAULT_ADDR` environment variable](https://github.com/hashicorp/vault/commit/f11c8febadc2a820d8ab76cae64d83adf60d59c7) (since almost the [beginning](https://github.com/hashicorp/vault/commit/cbf2d74670286f1152781d12a42e774ce692aa24) of the project).

**Comment by h3m5k** [comment]
I've tested injecting `ansible_` prefixed variables as _extra_vars:_ in a custom credential type with AWX 15.0.1, and this works just fine. 

Our use case is exactly what is described in https://github.com/ansible-collections/community.hashi_vault/issues/49#issuecomment-763802213 and the nice thing about this is that the only thing needed to provide the hashi_vault params is to attach the credential containing them, to any job template that needs them.
However, it is not entirely clear to me if this functionality will remain with the proposed `vars:` implementation? 
In other words, will we be able to switch the AWX credential injector config from this:
```
env:
  VAULT_AUTH_METHOD: '{{ hc_vault_auth_method }}'
  VAULT_ADDR: '{{ hc_vault_addr }}'
  VAULT_ROLE_ID: '{{ hc_vault_role_id }}'
  VAULT_SECRET_ID: '{{ hc_vault_secret_id }}'
```
...to this:
```
extra_vars:
  ansible_hashi_vault_auth: '{{ hc_vault_auth_method }}'
  ansible_hashi_vault_addr: '{{ hc_vault_addr }}'
  ansible_hashi_vault_rid: '{{ hc_vault_role_id }}'
  ansible_hashi_vault_sid: '{{ hc_vault_secret_id }}'
```
...and keep using hashi_vault without directly declaring any params, ie:
`lookup('community.hashi_vault.hashi_vault', 'secrets/data/mysecret')`


If this is the case, it would be nice to see this implemented prior to v2.0.0, and the removal of `VAULT_` variables, in order to keep using hashi_vault lookups without having to specify any parameters apart from _secret_.

If not, I don't see any real added value, from the AWX perspective, with adding _vars:_ to hashi_vault, as we already can inject whatever _extra_vars:_ we like, using custom credential types, and we would still have to use hashi_vault params where the lookup is used, which is what we want to avoid.

**Comment by briantist** [comment]
@h3m5k that's great! thank you so much for confirming that.

Just to be certain: did you inspect the values of the extra vars injected to ensure they really were available? I mainly ask that because of this: https://github.com/ansible/awx/pull/2363
Before that change, the credential injector did not prevent you from "using" `ANSIBLE_` prefixed env vars; it just didn't actually inject them at all; it silently failed. Want to be sure there's not something similar going on with injecting extra vars (and I don't have or use AWX, so I really appreciate the help here!)

---

Assuming all that works, then yes, inclusion of `vars:` entries for the options should allow you to continue using the lookup the way you expected.

> If this is the case, it would be nice to see this implemented prior to v2.0.0, and the removal of VAULT_ variables, in order to keep using hashi_vault lookups without having to specify any parameters apart from secret.

I completely agree. I will endeavor to have that implemented before (hopefully well before) any removal, in order to have a smooth path forward.

The main thing slowing it for the moment is that I'm trying to shore up testing on what's currently in the plugin, and avoid adding anything new without ensuring it will be thoroughly tested, and trying to be more thoughtful about naming of new vars/options, as it relates to the possibility of having more plugins in the collection.

If there's any doubt that this is important to me though: I was looking to add `vars:` entries before this issue came up, as even in my own production environment, we seek to not include most of those parameters in the `lookup()` call itself, and we currently do it with a homegrown wrapper around the main lookup that I'd very much like to get rid of 🙂. So I'll be looking to get something out sooner rather than later.

**Comment by h3m5k** [comment]
> Just to be certain: did you inspect the values of the extra vars injected to ensure they really were available? I mainly ask that because of this: [ansible/awx#2363](https://github.com/ansible/awx/pull/2363)

I did verify that the extra_vars were available. Twice now actually, since you made me unsure. :slightly_smiling_face:  

> Assuming all that works, then yes, inclusion of `vars:` entries for the options should allow you to continue using the lookup the way you expected.

Beatiful! Thanks for clarifying. :pray:

**Comment by briantist** [comment]
For anyone who might be watching this thread, I have a PR open that adds `vars:` entries.
I'd really appreciate if you have a chance to take a look and maybe test it out with AWX to see if it works for you!
👉 #86 👈

**Comment by briantist** [comment]
The changes referenced above have been [released in v1.2.0](https://github.com/ansible-collections/community.hashi_vault/releases/tag/1.2.0).
